tsfc/coffee_mode.py

```diff
@@ -93,69 +84,63 @@ def monomial_sum_to_expression(monomial_sum):
-            return
-        if key(solution) >= key(optimal_solution):
-            return
-        solution.add(variables[idx])
-        if is_feasible(solution):
-            if key(solution) < key(optimal_solution):
-                optimal_solution.clear()
-                optimal_solution.update(solution)
-            # No need to search further as adding more variables will
-            # only make the solution worse.
-        else:
-            solve(idx + 1)
-        solution.remove(variables[idx])
-        solve(idx + 1)
+def index_extent(factor, argument_indices):
+    """Compute the product of the extents of argument indices of a GEM expression
 
-    solve(0)
+    :arg factor: GEM expression
+    :arg argument_indices: set of argument indices
 
-    return optimal_solution
+    :returns: product of extents of argument indices
+    """
+    return numpy.prod([i.extent for i in factor.free_indices if i in argument_indices])
 
 
 def find_optimal_atomics(monomials, argument_indices):
     """Find optimal atomic common subexpressions, which produce least number of
     terms in the resultant IndexSum when factorised.
 
-    :arg monomials: An iterable of :class:`Monomial`s, all of which should have
+    :arg monomials: A list of :class:`Monomial`s, all of which should have
                     the same sum indices
     :arg argument_indices: tuple of argument indices
 
     :returns: list of atomic GEM expressions
     """
     atomics = tuple(OrderedDict.fromkeys(itertools.chain(*(monomial.atomics for monomial in monomials))))
 
-    def is_feasible(solution):
-        # Solution is only feasible if it intersects with all monomials
-        # Potentially can improve this by keeping track of violated constraints
-        # and suggest the next atomic to try (instead of just returning True or False)
-        return all(solution.intersection(monomial.atomics) for monomial in monomials)
-
     def cost(solution):
         extent = sum(map(lambda atomic: index_extent(atomic, argument_indices), solution))
         # Prefer shorter solutions, but larger extents
         return (len(solution), -extent)
 
-    optimal_atomics = solve_ip(atomics, is_feasible, key=cost)
-    return tuple(atomic for atomic in atomics if atomic in optimal_atomics)
+    optimal_solution = set(atomics)  # pessimal but feasible solution
+    solution = set()
+
+    max_it = 1 << 12
+    it = iter(range(max_it))
+
+    def solve(idx):
+        while idx < len(monomials) and solution.intersection(monomials[idx].atomics):
+            idx += 1
+
+        if idx < len(monomials):
+            if len(solution) < len(optimal_solution):
+                for atomic in monomials[idx].atomics:
+                    solution.add(atomic)
+                    solve(idx + 1)
+                    solution.remove(atomic)
+        else:
+            if cost(solution) < cost(optimal_solution):
+                optimal_solution.clear()
+                optimal_solution.update(solution)
```

Comment:
We reached the end of the list of monomials so we have a feasible solution.  If it is cheaper than the current one, then accept.